### ISSUE TITLE
"if" in a commented line parsed as an actual "if statement" in regards to matchit working with nvim-treesitter

### ISSUE THREAD

**Issue by comiluv** [neovim/neovim#15462]
### Neovim version (nvim -v)

NVIM v0.6.0-dev+142-g5a111c1b0

### Vim (not Nvim) behaves the same?

no, vim doesn't support nvim-treesitter

### Operating system/version

Microsoft Windows [Version 10.0.19043.1165]

### Terminal name/version

Windows Terminal/1.10.1933.0

### $TERM environment variable

(nothing is in $TERM env)

### Installation

choco install neovim --pre

### How to reproduce the issue

1. Install [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter)
2. Make a viml file: test-vim.vim
3. Type following
```viml
if
	" A comment containing a word if
else
	echo ELSE
endif
```
4. Press % while the cursor is on endif

### Expected behavior

Commented line is ignored for matching if-else-endif with % key.
![GIF 2021-08-22 오후 7-43-05](https://user-images.githubusercontent.com/19642705/130394191-feeb0654-fb68-4ea9-b8a1-bc13ce5272d6.gif)

### Actual behavior

The cursor jumps to the word if in the commented line
![GIF 2021-08-22 오후 7-42-23](https://user-images.githubusercontent.com/19642705/130394073-6c0a3dd8-83c8-4e02-89d0-5ff38ded2131.gif)

**Comment by comiluv** [comment]
This issue was posted on nvim-treesitter/nvim-treesitter#1732 but was discussed to be a problem with matchit

**Comment by vigoux** [comment]
Is a matchit issue when `syntax off`, thus unrelated to tree-sitter i'd say ?

**Comment by justinmk** [comment]
hopefully a (very small and simple) matchit replacement will materialize, which uses treesitter queries. Will keep this open to track that.

**Comment by timtyrrell** [comment]
@justinmk I know this exists: https://github.com/andymass/vim-matchup#tree-sitter-integration

**Comment by kaddkaka** [comment]
This behavior for `%` is surprising and manifests as a regression when a user activates treesitter.

Would it be possible to incorporate something like https://github.com/andymass/vim-matchup into neovim core to be more robust when someone activates treesitter?

**Comment by clason** [comment]
That very much depends on the "something". A minimal(!) Lua alternative to matchit that is treesitter-aware is welcome.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/31563 adds some treesitter capability to matchit.

Keeping this issue open to track a full Lua implementation/rewrite of matchit.

**Comment by yochem** [comment]
Concerning the Lua rewrite:

- > it should not require its own set of queries but piggy-back on the existing highlight queries

These differ from the matchit matches. E.g. in lua code, matchit cycles between if/elseif/else/end, but the highlight query for if statements also includes "then". Would this be a problem?

- Does it also need to support syntax/non-treesitter buffers? Or would it be treesitter-only?

**Comment by clason** [comment]
> These differ from the matchit matches. E.g. in lua code, matchit cycles between if/elseif/else/end, but the highlight query for if statements also includes "then". Would this be a problem?

Not necessarily. Matching (heh!) behavior 1:1 is not a goal. (And there are always more fully featured and customizable plugins like vim-matchup which this is not meant to replace.)

> Does it also need to support syntax/non-treesitter buffers?

Yes, it should be a full replacement so we can replace matchit/matchparen entirely.
